### PR TITLE
[BEAM-858] Add ApexRunner integration test.

### PR DESCRIPTION
**Pull request by tweise**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:
- [ ] Make sure the PR title is formatted like:
  `[BEAM-<Jira issue #>] Description of pull request`
- [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
     Travis-CI on your fork and ensure the whole test matrix passes).
- [ ] Replace `<Jira issue #>` in the title with the actual Jira issue
     number, if there is one.
- [ ] If this contribution is large, please file an Apache
     [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).
---

R: @kennknowles

### CODE REVIEW

**Comment by tweise** [comment]
There is a conflict with a property in ApexPipelineOptions:

```
Caused by: java.lang.IllegalArgumentException: methods with same signature getParallelism() but incompatible return types: int and others
	at sun.misc.ProxyGenerator.checkReturnTypes(ProxyGenerator.java:656)
	at sun.misc.ProxyGenerator.generateClassFile(ProxyGenerator.java:461)
	at sun.misc.ProxyGenerator.generateProxyClass(ProxyGenerator.java:339)
	at java.lang.reflect.Proxy$ProxyClassFactory.apply(Proxy.java:639)
	at java.lang.reflect.Proxy$ProxyClassFactory.apply(Proxy.java:557)
	at java.lang.reflect.WeakCache$Factory.get(WeakCache.java:230)
	at java.lang.reflect.WeakCache.get(WeakCache.java:127)
	at java.lang.reflect.Proxy.getProxyClass0(Proxy.java:419)
	at java.lang.reflect.Proxy.getProxyClass(Proxy.java:371)
	at org.apache.beam.sdk.options.PipelineOptionsFactory.validateWellFormed(PipelineOptionsFactory.java:623)
	at org.apache.beam.sdk.options.PipelineOptionsFactory.register(PipelineOptionsFactory.java:561)
	at org.apache.beam.sdk.options.PipelineOptionsFactory.initializeRegistry(PipelineOptionsFactory.java:587)
	at org.apache.beam.sdk.options.PipelineOptionsFactory.<clinit>(PipelineOptionsFactory.java:536)
	... 37 more
```
 Not sure why this comes up here only. I have commented out the property temporarily.

**Comment by kennknowles** [comment]
The other declaration is in `FlinkPipelineOptions` and they have to match exactly. I think switching `int` to `Integer` should fix it. The defaults might need to match too, I cannot remember.

**Comment by kennknowles** [comment]
This error message is poor. I think it is worth a JIRA to improve it and catch the error in our own validation, not the ProxyGenerator.

**Comment by tweise** [comment]
The tests pass in the last iteration, squashed the commits. I need to work on how to determine from the source that there is no more input so the runner can terminate (vs. sitting idle till timeout). Here the source is `TextIO.Read.from(options.getInputFile())`. When the source is bounded and advance returns false, can the pipeline shutdown. Currently that's only done for ValuesSource:

https://github.com/apache/incubator-beam/blob/apex-runner/runners/apex/src/main/java/org/apache/beam/runners/apex/translators/io/ApexReadUnboundedInputOperator.java#L77

**Comment by tweise** [comment]
@kennknowles PR ready or review.

**Comment by kennknowles** [comment]
Once `advance` returns false, you actually _must_ close it: https://github.com/apache/incubator-beam/blob/master/sdks/java/core/src/main/java/org/apache/beam/sdk/io/BoundedSource.java#L89

For unbounded data, when the watermark reaches infinity (`BoundedWindow.TIMESTAMP_MAX_VALUE`) then you can/should shut it down. Only some runners manage to do this.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/ApexPipelineOptions.java [hunk 1]
```diff
@@ -32,12 +32,6 @@
 
   String getApplicationName();
 
-  @Description("set parallelism for Apex runner")
-  void setParallelism(int parallelism);
```
Comment (kennknowles) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/ApexPipelineOptions.java]: I think you do want to keep this, don't you? It is unfortunate that we have this global options namespace, but that is all about the ability to make it command-line friendly. I think perhaps you should match the Flink value exactly, or (better) factor this into some common `Options` class.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java [hunk 2]
```diff
@@ -139,12 +140,8 @@ public void populateDAG(DAG dag, Configuration conf) {
...
       if (options.getRunMillis() > 0) {
         try {
-          long timeout = System.currentTimeMillis() + options.getRunMillis();
-          while (System.currentTimeMillis() < timeout) {
-            if (assertionError != null) {
-              throw assertionError;
-            }
-          }
+          // this is necessary for tests that just call run() and not waitUntilFinish
+          ApexRunnerResult.waitUntilFinished(lc, Duration.millis(options.getRunMillis()));
```
Comment (kennknowles) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java]: I think you should do this in `TestApexRunner` only. Eventually, `TestPipeline` should own this. Maybe that can happen even right now?

File: runners/apex/src/test/java/org/apache/beam/runners/apex/examples/StreamingWordCountTest.java [hunk 3]
```diff
@@ -91,7 +91,7 @@ public void testWindowedWordCount() throws Exception {
     ApexPipelineOptions options = PipelineOptionsFactory.fromArgs(args).withValidation()
         .as(ApexPipelineOptions.class);
     options.setApplicationName("StreamingWordCount");
-    options.setParallelism(1);
+    //options.setParallelism(1);
```
Comment (kennknowles) [inline runners/apex/src/test/java/org/apache/beam/runners/apex/examples/StreamingWordCountTest.java]: Even with the option available, you shouldn't do this, should you? (I didn't read the surrounding code that much - maybe a comment about why you want it to be `1`?)

File: runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java (hunk 2 above)
Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/ApexRunner.java]: runMillis is only set in the test runner. And yes, I think this can be now be moved over to the test runner, should I do it as part of this PR?

File: runners/apex/src/test/java/org/apache/beam/runners/apex/examples/StreamingWordCountTest.java (hunk 3 above)
Comment (tweise) [inline runners/apex/src/test/java/org/apache/beam/runners/apex/examples/StreamingWordCountTest.java]: Yes, I will remove the line.

**Comment by tweise** [comment]
The fix for BoundedSource is in #1231

I think I will merge those changes into this PR.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/ApexPipelineOptions.java (hunk 1 above)
Comment (kennknowles) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/ApexPipelineOptions.java]: This might make an interesting discussion, depending on what `parallelism` means here. Each runner handles this differently, and the Beam model quite clearly leaves it undefined so runners can do their own thing. I don't think `PipelineOptions` can (or should) take care of it with that level of detail.

Since it is BEAM-838, no need to solve it here.

**Comment by tweise** [comment]
@kennknowles done adding the changes from other PR

File: runners/apex/src/main/java/org/apache/beam/runners/apex/examples/WordCount.java [hunk 4]
```diff
@@ -0,0 +1,106 @@
...
+import org.apache.beam.sdk.transforms.windowing.FixedWindows;
+import org.apache.beam.sdk.transforms.windowing.Window;
+import org.apache.beam.sdk.values.KV;
+import org.joda.time.Duration;
+
+
+/**
+ * Word count example on Apex runner.
+ */
+public class WordCount {
```
Comment (kennknowles) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/examples/WordCount.java]: Why did you need to add this? You should be able to use the common `WordCount` example. I don't see anything obviously different. The other runner-specific `examples` directories really need to be deleted, actually.

File: runners/apex/src/test/java/org/apache/beam/runners/apex/examples/StreamingWordCountTest.java [hunk 5]
```diff
@@ -110,12 +113,26 @@ public void testWindowedWordCount() throws Exception {
...
     result.cancel();
     Assert.assertTrue(
         FormatAsStringFn.RESULTS.containsKey("foo") && FormatAsStringFn.RESULTS.containsKey("bar"));
     FormatAsStringFn.RESULTS.clear();
 
   }
+
+  @Ignore
+  @Test
+  public void testWordCountExample() throws Exception {
```
Comment (kennknowles) [inline runners/apex/src/test/java/org/apache/beam/runners/apex/examples/StreamingWordCountTest.java]: Same here. You should be able to use the existing `WordCountIT`.

Comment (tweise) [inline runners/apex/src/test/java/org/apache/beam/runners/apex/examples/StreamingWordCountTest.java]: It would be a circular dependency. Examples depend on runner, runner depends on examples. To get there, the example pipelines should be separate from the integration test project.

File: runners/apex/src/main/java/org/apache/beam/runners/apex/examples/WordCount.java (hunk 4 above)
Comment (tweise) [inline runners/apex/src/main/java/org/apache/beam/runners/apex/examples/WordCount.java]: By adding it (would be glad to reuse if possible) I can write the test that already uncovered 3 issues that went unnoticed before.

**Comment by tweise** [comment]
@kennknowles I removed the example from this PR, will take it up as part of BEAM-844.
